api/v1alpha4/clusterclass_webhooks.go

```diff
@@ -0,0 +1,154 @@
+	oldClusterClass, ok := old.(*ClusterClass)
+	if !ok {
+		return apierrors.NewBadRequest(fmt.Sprintf("expected a ClusterClass but got a %T", old))
+	}
+	return in.validate(oldClusterClass)
+}
+
+// ValidateDelete implements webhook.Validator so a webhook will be registered for the type.
+func (in *ClusterClass) ValidateDelete() error {
+	return nil
+}
+
+func (in *ClusterClass) validate(old *ClusterClass) error {
+	var allErrs field.ErrorList
+
+	// ensure all the references are within the same namespace
+	if in.Spec.Infrastructure.Ref.Namespace != in.Namespace {
+		allErrs = append(
+			allErrs, field.Invalid(field.NewPath("spec", "infrastructure", "ref", "namespace"), in.Spec.Infrastructure.Ref.Namespace, "must match metadata.namespace"),
+		)
+	}
+	if in.Spec.ControlPlane.Ref.Namespace != in.Namespace {
+		allErrs = append(
+			allErrs, field.Invalid(field.NewPath("spec", "controlPlane", "ref", "namespace"), in.Spec.ControlPlane.Ref.Namespace, "must match metadata.namespace"),
+		)
+	}
+	for _, class := range in.Spec.Worker.Deployments {
+		if class.Template.Bootstrap.Ref.Namespace != in.Namespace {
+			allErrs = append(allErrs,
+				field.Invalid(field.NewPath("ClusterClass", "spec", "worker", "deployments", "template", "bootstrap", "ref", "namespace"), class.Template.Bootstrap.Ref.Namespace, "must match metadata.namespace"),
+			)
+		}
+		if class.Template.Infrastructure.Ref.Namespace != in.Namespace {
+			allErrs = append(allErrs,
+				field.Invalid(field.NewPath("ClusterClass", "spec", "worker", "deployments", "infrastructure", "bootstrap", "ref", "namespace"), class.Template.Infrastructure.Ref.Namespace, "must match metadata.namespace"),
+			)
+		}
+	}
+
+	// Ensure MachineDeployment class are unique.
+	classNames := make(map[string]struct{})
+	for _, class := range in.Spec.Worker.Deployments {
+		if _, ok := classNames[class.Class]; ok {
+			allErrs = append(allErrs,
+				field.Invalid(field.NewPath("ClusterClass", "spec", "worker", "deployments"), class, fmt.Sprintf("MachineDeployment class %q is defined more than once.", class.Class)),
+			)
+		}
+		classNames[class.Class] = struct{}{}
+	}
+
+	// In case of updates
+	if old != nil {
+		// Makes sure all the old MachineDeployment class are still there (only MachineDeployment class addition are allowed).
+		oldClassNames := make(map[string]struct{})
+		for _, oldClass := range old.Spec.Worker.Deployments {
+			if _, ok := classNames[oldClass.Class]; !ok {
+				allErrs = append(allErrs,
+					field.Invalid(field.NewPath("ClusterClass", "spec", "worker", "deployments"), in.Spec.Worker.Deployments, fmt.Sprintf("The %q MachineDeployment class can't be removed.", oldClass.Class)),
+				)
+			}
+			oldClassNames[oldClass.Class] = struct{}{}
+		}
+
+		// Makes sure no additional changes were applied.
+		// NOTE: in order to do so, we are considering only the Deployments existing both in the old and new cluster class;
+		tmpIn := in.DeepCopy()
```

Comment:
Why do we need a copy?

Reply:
We need a copy because we are required to remove from the objects the new MachineDeployments in order to compare with old when checking for mutations not yet supported.